### ISSUE TITLE
Matchers in post request body and post request headers

### ISSUE THREAD

**Issue by soundstep** [pact-foundation/pact-mock_service#57]
Hi,
I can't seem to be able to use ```somethingLike``` matchers in a POST request (body and headers).
Is that normal? Am I missing something?
See this gist:
https://gist.github.com/soundstep/4991f67b98d274efbe18ef4eb0c53baa
The only ```matchingRules``` created are for the body of the response.
FYI: I originally reported this issue on the js implementation: https://github.com/pact-foundation/pact-js/issues/16
Any help appreciated.
Cheers.

**Comment by bethesque** [comment]
I believe this will answer your question. Let me know if it doesn't. https://github.com/bethesque/pact-mock_service/issues/55

In summary, the request matching info is used by the mock service, but not serialised to the pact, as it is not needed in the verification step. When verifying a pact, you need a concrete examples - the fuzzy matching logic from when it was used in the consumer is irrelevant.

**Comment by soundstep** [comment]
I understand that for a path. But my example is a generated id that goes in the body of a post request.
This is a very common thing to do, especially for authentication.

So it is all fine to create the pact file and send it to the broker.
However, in my example, the ```nonce``` is a generate id that serves the purpose of identifying auth request. It will be unique per request and per client.
This makes me unable to use the pack-mock-service to authenticate from the client due to this generated id.

These matchers have been created to solved a problem when servers generates something that can be different for each request (dates, tokens, ids). This is also true for the client: dates, tokens, ids can be generated on the client side.

Can you advice how I can address this issue of having a "request id" per request and still be able to use the pack-mock-service for development/workflow?

This is what I get from trying to authenticate from the client:

```
{
    "message": "No interaction found for POST /auth",
    "interaction_diffs": [
        {
            "description": "a login request for a content token",
            "provider_state": "User ctv.integration.test@example.com exists",
            "body": {
                "nonce": {
                    "EXPECTED": "abc123",
                    "ACTUAL": "GENERATED_ID"
                }
            }
        }
    ]
}```

**Comment by soundstep** [comment]
Something to add is that, this is not only a development/workflow issue, but also a CI issue.
We were planning to also use the same pack-mock-service for acceptance tests (selenium ran tests), fed with the interactions posted to the broker from the unit tests.

**Comment by bethesque** [comment]
I'm unclear as to which step you are having problems with - running the tests in the consumer, or verifying the pact on the provider. I've run the gist you provided, and it runs fine for me.

```

issue57 $ pact-mock-service start
[2016-11-15 10:49:14] INFO  WEBrick 1.3.1
[2016-11-15 10:49:14] INFO  ruby 2.0.0 (2013-11-22) [x86_64-darwin13.0.0]
[2016-11-15 10:49:14] INFO  WEBrick::HTTPServer#start: pid=15714 port=1234
issue57 $ ./pact-test.sh
I, [2016-11-15T10:49:20.568160 #15714]  INFO -- : Cleared interactions before example ""
Deleted interactions

I, [2016-11-15T10:49:21.587284 #15714]  INFO -- : Registered expected interaction POST /auth
D, [2016-11-15T10:49:21.588255 #15714] DEBUG -- : {
  "description": "duh",
  "request": {
    "method": "POST",
    "path": "/auth",
    "headers": {
      "Content-Type": "application/json",
      "Accept": "application/json"
    },
    "body": {
      "nonce": {
        "json_class": "Pact::SomethingLike",
        "contents": "abc123"
      },
      "username": "ctv.integration.test@example.com",
      "scope": "content",
      "grant_type": "password",
      "password": "pw123456"
    }
  },
  "response": {
    "status": 200,
    "headers": {
      "Content-Type": "application/json; charset=UTF-8",
      "Access-Control-Allow-Origin": "*",
      "Cache-Control": "no-store"
    },
    "body": {
      "access_token": {
        "json_class": "Pact::SomethingLike",
        "contents": "eyJhbGc"
      },
      "token_type": "bearer",
      "refresh_token": {
        "json_class": "Pact::SomethingLike",
        "contents": "eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwcz"
      }
    }
  }
}
Set interactionsI, [2016-11-15T10:49:21.607751 #15714]  INFO -- : Registered expected interaction OPTIONS /auth
D, [2016-11-15T10:49:21.608050 #15714] DEBUG -- : {
  "description": "duh preflight",
  "request": {
    "method": "OPTIONS",
    "path": "/auth",
    "headers": {
      "Access-Control-Request-Headers": {
        "json_class": "Pact::SomethingLike",
        "contents": "content-type"
      }
    }
  },
  "response": {
    "status": 200,
    "headers": {
      "Access-Control-Allow-Methods": "OPTIONS, GET",
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Headers": "content-type"
    }
  }
}
Set interactions

I, [2016-11-15T10:49:21.625031 #15714]  INFO -- : Received request OPTIONS /auth
D, [2016-11-15T10:49:21.625415 #15714] DEBUG -- : {
  "path": "/auth",
  "query": "",
  "method": "options",
  "headers": {
    "Host": "localhost:1234",
    "Pragma": "no-cache",
    "Access-Control-Request-Method": "POST",
    "Origin": "http://localhost:8001",
    "Accept-Encoding": "gzip, deflate, sdch, br",
    "Accept-Language": "en-GB,en-US;q=0.8,en;q=0.6,fr;q=0.4,ms;q=0.2",
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36",
    "Accept": "*/*",
    "Cache-Control": "no-cache",
    "Referer": "http://localhost:8001/debug.html?bypassLoading=1&debug=1",
    "Connection": "keep-alive",
    "Access-Control-Request-Headers": "content-type",
    "Version": "HTTP/1.1"
  }
}
I, [2016-11-15T10:49:21.647604 #15714]  INFO -- : Found matching response for OPTIONS /auth
D, [2016-11-15T10:49:21.647985 #15714] DEBUG -- : {
  "status": 200,
  "headers": {
    "Access-Control-Allow-Methods": "OPTIONS, GET",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "content-type"
  }
}
I, [2016-11-15T10:49:21.664641 #15714]  INFO -- : Received request POST /auth
D, [2016-11-15T10:49:21.664898 #15714] DEBUG -- : {
  "path": "/auth",
  "query": "",
  "method": "post",
  "body": {
    "nonce": "abc123",
    "username": "ctv.integration.test@example.com",
    "scope": "content",
    "grant_type": "password",
    "password": "pw123456"
  },
  "headers": {
    "Content-Length": "128",
    "Content-Type": "application/json",
    "Host": "localhost:1234",
    "Pragma": "no-cache",
    "Origin": "http://localhost:8001",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-GB,en-US;q=0.8,en;q=0.6,fr;q=0.4,ms;q=0.2",
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36",
    "Accept": "application/json",
    "Cache-Control": "no-cache",
    "Referer": "http://localhost:8001/debug.html?bypassLoading=1&debug=1",
    "Connection": "keep-alive",
    "Version": "HTTP/1.1"
  }
}
I, [2016-11-15T10:49:21.665718 #15714]  INFO -- : Found matching response for POST /auth
D, [2016-11-15T10:49:21.666022 #15714] DEBUG -- : {
  "status": 200,
  "headers": {
    "Content-Type": "application/json; charset=UTF-8",
    "Access-Control-Allow-Origin": "*",
    "Cache-Control": "no-store"
  },
  "body": {
    "access_token": {
      "json_class": "Pact::SomethingLike",
      "contents": "eyJhbGc"
    },
    "token_type": "bearer",
    "refresh_token": {
      "json_class": "Pact::SomethingLike",
      "contents": "eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwcz"
    }
  }
}
{"access_token":"eyJhbGc","token_type":"bearer","refresh_token":"eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwcz"}

I, [2016-11-15T10:49:21.682698 #15714]  INFO -- : Verifying - interactions matched for example ""
Interactions matched

I, [2016-11-15T10:49:21.700377 #15714]  INFO -- : Writing pact with details {:consumer=>{:name=>"Consumer Name"}, :provider=>{:name=>"Provider Name"}, :pact_dir=>"pacts", :pact_specification_version=>"2"}
I, [2016-11-15T10:49:21.700580 #15714]  INFO -- : Writing pact for Provider Name to pacts/consumer_name-provider_name.json
{
  "consumer": {
    "name": "Consumer Name"
  },
  "provider": {
    "name": "Provider Name"
  },
  "interactions": [
    {
      "description": "duh preflight",
      "request": {
        "method": "OPTIONS",
        "path": "/auth",
        "headers": {
          "Access-Control-Request-Headers": "content-type"
        }
      },
      "response": {
        "status": 200,
        "headers": {
          "Access-Control-Allow-Methods": "OPTIONS, GET",
          "Access-Control-Allow-Origin": "*",
          "Access-Control-Allow-Headers": "content-type"
        }
      }
    },
    {
      "description": "duh",
      "request": {
        "method": "POST",
        "path": "/auth",
        "headers": {
          "Content-Type": "application/json",
          "Accept": "application/json"
        },
        "body": {
          "nonce": "abc123",
          "username": "ctv.integration.test@example.com",
          "scope": "content",
          "grant_type": "password",
          "password": "pw123456"
        }
      },
      "response": {
        "status": 200,
        "headers": {
          "Content-Type": "application/json; charset=UTF-8",
          "Access-Control-Allow-Origin": "*",
          "Cache-Control": "no-store"
        },
        "body": {
          "access_token": "eyJhbGc",
          "token_type": "bearer",
          "refresh_token": "eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwcz"
        },
        "matchingRules": {
          "$.body.access_token": {
            "match": "type"
          },
          "$.body.refresh_token": {
            "match": "type"
          }
        }
      }
    }
  ],
  "metadata": {
    "pactSpecificationVersion": "2.0.0"
  }
}
```

**Comment by bethesque** [comment]
That was with pact-mock_service-0.10.2

**Comment by soundstep** [comment]
Yes the gist run fine. Creating the json pact file is fine (unit test phase), sending it to the broker is fine, verifying the pact on the server is also fine.
The thing I'm unable to do is use the pack mock service for our acceptance tests in our CI pipeline.

The workflow in our CI is:
1. start pact mock service
2. feed the service with the interactions from the server
3. launch the client using selenium
4. use the real client to make the request to the pact mock service.

The request will fail against the pact mock service because one of the property in the body of the post request is dynamic and generated: ```body.nonce```.

![screen-shot-2016-11-14-at-11 26 20](https://cloud.githubusercontent.com/assets/133293/20298615/c1ba28be-ab0e-11e6-9ce7-a84974b686e3.png)

I was expecting using a "something like" matcher would also work for the body of the request, as it does for the body of the response (we have tokens generated in the response).

You might consider that as a feature request more than an issue.

This pushes me to expose this ```nonce``` parameters via a config so that the property can be hard-coded. Fortunately there is no security issue doing so, even though it is still a very bad practice, but if there was a security concern with this parameters I would simply be unable to use the pact mock service.

Hope that helps.
Let me know what you think.

Cheers.

**Comment by bethesque** [comment]
Are you setting up the interactions in the mock server using the interactions that were written in the pact file?

**Comment by soundstep** [comment]
Yes exactly.
The pact file is created and sent to the server at the unit tests level.

**Comment by bethesque** [comment]
Oh right. _Now_ I see your problem. Yeah, that won't work (as you've noticed!) as the matching info is discarded when the pact is written. That's not how a pact is designed to be used. Even if the matching rules were serialised to the pact, it still wouldn't work because the format for setting up interactions on the mock server is different from the pact format (not deliberately, just because of historical reasons).

You'll need to load the interactions into the mock server from the same source as the unit tests. I can show you how to do that using shared fixtures. Are your CI tests run in ruby?

**Comment by soundstep** [comment]
Unit tests are javascript, and acceptance tests are ruby with selenium, in different dockers environments.

I tried what you said but it is not even working with a curl.

Using the application directly (selenium) alters the request as the ```body.nonce``` in the post request is generated and always different.

A curl to the mock service with a body of:

```
{
    nonce: "abc123"
}
```

would work because the interactions has been set with: 

```
withRequest: {
    method: 'post',
    path: '/auth',
    body: {
        nonce: Pact.Matchers.somethingLike('abc123')
```

However, using the application in the selenium tests creates a request with a body of:

```
{
    nonce: "generated id by client"
}
```

Which makes the pact mock service answer with a "no interaction found".

So basically you are saying that pact has not been designed to handled this kind of issues?
Which means that the only workaround for me is to exposed this parameters so it can hard coded in tests where the application is used?

**Comment by soundstep** [comment]
I made another test by keeping the same server running and what you said does work, the server seems to hold the "something like matcher".
I will make more tests. I keep you posted, thanks for your help!

**Comment by soundstep** [comment]
So I confirm if I use the same server the matcher works, and also if I use 2 different servers with the same fixture.

The breaking point is the json output, if I use the json output to add interactions to the mock service, the matchers are not there.

They seem to say that it should work there:
https://github.com/realestate-com-au/pact/wiki/Regular-expressions-and-type-matching-with-Pact

```
You can embed Pact::Terms or Pact::SomethingLike inside the hash. Remember that all query params will be parsed to strings, so don't use a SomethingLike with a number.
```

So tell me if I'm wrong, because I don;t the pact specification very well.
It seems matchers in bodies are just not implemented in the pact-mock service you built?
And the matchers should be part of the json output sent to the broker?

If it is not implemented on your side but would still be valid to include "body matching rules" in the json against the pact specification, would you consider a pull request?

Please let me know!

**Comment by bethesque** [comment]
> The breaking point is the json output, if I use the json output to add interactions to the mock service, the matchers are not there.

This is the crux. You can't reload a pact into a mock server. It's not designed to work that way. You have to use the source interaction JSON each time.

> It seems matchers in bodies are just not implemented in the pact-mock service you built?

The matchers are indeed implemented for request bodies, it's just that the pact that you are loading does not have the matchers in it.

> And the matchers should be part of the json output sent to the broker?

Only the response matchers are written to the pact. The request matchers aren't needed in the verification process, so they are discarded. 

> would you consider a pull request

I will have a look at the code and work out what changes would need to be made to support reloading a pact into the mock server. Keep in mind, _it's not designed to work like this_. If you load all the interactions from a pact at one time into the mock server, you won't be able to test the same endpoint with different scenarios (ie. different response bodies or status codes) because it won't know which response to return.

**Comment by bethesque** [comment]
I suspect this might work, but I don't have time to test it right now. If you're confident enough to hack into your gem file and overwrite `lib/pact/consumer_contract/request_decorator.rb` with the following code, you could see if the matching rules for the request get written into the pact.

```ruby
require 'pact/reification'

module Pact
  class RequestDecorator

    def initialize request, decorator_options = {}
      @request = request
      @decorator_options = decorator_options
    end

    def to_json(options = {})
      as_json.to_json(options)
    end

    def as_json options = {}
      hash = {
        method: request.method,
        path: path
      }
      hash[:query]   = query   if request.specified?(:query)
      hash[:headers] = headers if request.specified?(:headers)
      hash[:body]    = body    if request.specified?(:body)
      include_matching_rules? ? with_matching_rules(hash) : hash
    end

    private

    attr_reader :request

    def path
      Pact::Reification.from_term(request.path)
    end

    def headers
      Pact::Reification.from_term(request.headers)
    end

    # This feels wrong to be checking the class type of the Query
    # Do this better somehow.
    def query
      Pact::Reification.from_term(request.query)
    end

    # This feels wrong to be checking the class type of the body
    # Do this better somehow.
    def body
      if content_type_is_form && request.body.is_a?(Hash)
        URI.encode_www_form convert_hash_body_to_array_of_arrays
      else
        Pact::Reification.from_term(request.body)
      end
    end

    def content_type_is_form
      request.content_type? 'application/x-www-form-urlencoded'
    end

    #This probably belongs somewhere else.
    def convert_hash_body_to_array_of_arrays
      arrays = []
      request.body.keys.each do | key |
        [*request.body[key]].each do | value |
          arrays << [key, value]
        end
      end
      Pact::Reification.from_term arrays
    end

    def include_matching_rules?
      pact_specification_version && !pact_specification_version.start_with?('1')
    end

    def with_matching_rules hash
      matching_rules = Pact::MatchingRules.extract hash
      example = Pact::Reification.from_term hash
      return example if matching_rules.empty?
      example.merge(matchingRules: matching_rules)
    end

    def pact_specification_version
      @decorator_options[:pact_specification_version]
    end

  end
end

```

**Comment by soundstep** [comment]
Thank you Beth, that's great help.
We will have a look to this in the next days, see how it goes, and come back to you!
Cheers.

**Comment by MattTYXM** [comment]
Has anyone taken care of this yet?

**Comment by bethesque** [comment]
Don't think so.

**Comment by soundstep** [comment]
Sorry still thinking about that one.
I haven't tested the code you sent yet, but I'll be back on pact stuff next week, I'll try to check with @itsDanOades and see if it makes sense.

**Comment by MattTYXM** [comment]
I have created a pull request (https://github.com/bethesque/pact-mock_service/pull/62) that I believe will solve the issues described here.

**Comment by soundstep** [comment]
Thanks @MattTYXM, I think @bethesque is still thinking about that one ;)
Did you have some real case were you also needed that capability?
Are we still confirm with the pact philosophy here?

Also how does this PR works with query? As query parameters are written that way in the pact file (a string and not an object):

```
"interactions": [
  {
    "request": {
      "query": "sortBy=title&features=hls&date=01-01-1970"
```

**Comment by bethesque** [comment]
Hey guys, sorry, I've just moved house, so I have almost no free time at the moment to look at code. I apologise for the delay.

**Comment by MattTYXM** [comment]
@soundstep I did run into the issue with the request query as an object. I would have really liked the ability to give each query param a term but at this point I just worked around it using regex. I took some time and fixed the query object issue but then the pact-jvm provider test was expecting query to be a string so at the end of the day I abandoned the effort. I can post the code that got me what you were asking for.

`
def query
      if include_matching_rules?
        request.query.query.each do | key, val |
          if val.length == 1
            request.query.query[key] = val[0]            
          end
        end

        request.query.query
      else
        Pact::Reification.from_term(request.query)
      end
    end
`
